cpp/src/arrow/compute/kernels/row_encoder_internal.h

```diff
@@ -424,49 +469,228 @@ struct ListFixedWidthChildEncoder : KeyEncoder {
+          bit_util::GetBit(list_validity_buffer, j)) {
+          Offset element_begin = list_offset_buffer[j];
+          Offset element_end = list_offset_buffer[j + 1];
+          Offset number_elements = element_end - element_begin;
+          length_sum += number_elements;
+        }
+      }
+      int32_t buffer_size = kExtraByteForNull + sizeof(Offset)
+                            + (kExtraByteForNull + sizeof(ChildOffset)) * list_size
+                            + length_sum;
+      return buffer_size;
+    };
+
+    if (data.is_array()) {
+      int64_t i = 0;
+      IterateListArray<T>(
+        data.array,
+        [&](Offset list_begin, Offset list_end, const uint8_t* list_validity_buffer,
+          const uint8_t* list_offset_buffer, const uint8_t*) {
+          lengths[i++] += getListBufferSize(list_begin, list_end,
+                                            list_validity_buffer, list_offset_buffer);
+        },
+        [&] { lengths[i++] += kExtraByteForNull + sizeof(Offset);});
+    } else {
+      ProcessListScalar<T>(
+        data.scalar_as<ScalarType>(),
+        [&](int64_t list_length, const uint8_t* scalar_validity_buffer,
+            const uint8_t* scalar_offset_buffer, const uint8_t*) {
+          scalar_buffer_length = getListBufferSize(0, list_length, scalar_validity_buffer,
+                                          scalar_offset_buffer);});
+      for (int64_t i = 0; i < batch_length; i++) {
+        lengths[i] += scalar_buffer_length;
+      }
+    }
+  }
+
+  void AddLengthNull(int32_t* length) override {
+    *length += kExtraByteForNull + sizeof(Offset);
+  }
+
+  Status Encode(const ExecValue& data, int64_t batch_length,
+                uint8_t** encoded_bytes) override {
+    auto encodeList = [&](Offset list_begin, Offset list_end,
+      const uint8_t* list_validity_buffer, const uint8_t* list_offset_buffer_,
+      const uint8_t* list_data_buffer, uint8_t* encoded_ptr) {
+      uint8_t* encoded_ptr_start = encoded_ptr;
+      EncodeVarLengthHeader<Offset>(encoded_ptr, kValidByte, list_end - list_begin);
+      const ChildOffset* list_offset_buffer =
+        reinterpret_cast<const ChildOffset*>(list_offset_buffer_);
+      for (Offset j = list_begin; j < list_end; j++) {
+        if (list_validity_buffer == NULLPTR ||
+          bit_util::GetBit(list_validity_buffer, j)) {
+          ChildOffset element_begin = list_offset_buffer[j];
+          ChildOffset element_end = list_offset_buffer[j + 1];
+          ChildOffset element_size = element_end - element_begin;
+          EncodeVarLengthElement<ChildOffset>(encoded_ptr, kValidByte, element_size,
+            std::string_view(reinterpret_cast<const char*>(
+            list_data_buffer + element_begin), element_size));
+          // EncodeVarLengthHeader<ChildOffset>(encoded_ptr, kValidByte, element_size);
+          // memcpy(encoded_ptr, list_data_buffer + element_begin, element_size);
+          // encoded_ptr += element_size;
```

Comment:
The commented code provides the same result as EncodeVarLengthElement in the end. I slightly prefer the uncommented code but I'm not sure how much faster it is to not construct a string_view.